woocommerce/woocommerce-ios
# [Woo POS] Implement reader update failure views - retryable, non-retryable, low battery

Closes: #13070 
Currently based off of https://github.com/woocommerce/woocommerce-ios/pull/13085



## Description


This PR implements the SwiftUI views for 3 card reader software update failure cases: retryable, non-retryable, low battery. For some reason, all the `.updatingFailed(tryAgain:...)` instances in the app set the `tryAgain` parameter to `nil`, perhaps we could remove this parameter if we don't plan to implement reader update retry in the foreseeable future.

## Testing information


For easier testing, feel free to mock the error by replacing `CardReaderConnectionController.onInitialization` with

* Non-retryable

```swift
        onUpdateFailed(error: CardReaderServiceError.softwareUpdate(underlyingError: .readerSoftwareUpdateFailed, batteryLevel: 0.2))
```

* Retryable

In addition to the replacement as in the non-retryable case above, also replace the `tryAgain` value to an empty closure:

https://github.com/woocommerce/woocommerce-ios/blob/47e1d4eebb7e5b8ff6c4aee41779538d0e3595c1/WooCommerce/Classes/ViewRelated/CardPresentPayments/CardReaderConnectionController.swift#L615

* Low battery error case

```swift
        onUpdateFailed(error: CardReaderServiceError.softwareUpdate(underlyingError: .readerSoftwareUpdateFailedBatteryLow, batteryLevel: 0.3))
```

And then:

Prerequisite: having a self-hosted store eligible for POS

For each of the 3 error cases with the testing setup above:

- Switch to a store in the prerequisite if needed
- Turn on the feature switch in Menu > Settings > Experimental Features > POS, if needed
- Go to the Menu tab --> the POS row should appear shortly
- Tap to enter POS
- Tap `Connect now` in the bottom bar --> the error modal should be shown

## Screenshots


retryable | non-retryable | low battery
-- | -- | --
![Simulator Screenshot - iPad (10th generation) - 2024-06-18 at 14 38 31](https://github.com/woocommerce/woocommerce-ios/assets/1945542/e1c82123-ad9e-4e52-af42-53c82e5214eb) | ![Simulator Screenshot - iPad (10th generation) - 2024-06-18 at 14 34 52](https://github.com/woocommerce/woocommerce-ios/assets/1945542/22d34f36-7cc3-4fb5-8170-cfa4236ccbc0) | ![Simulator Screenshot - iPad (10th generation) - 2024-06-18 at 14 06 55](https://github.com/woocommerce/woocommerce-ios/assets/1945542/dbd6afd0-c66a-4f78-bd41-0e7f40d1a9cc)


---
